File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.cpp
#include <string>
#include <vector>
#include <set>
#include <unordered_map>

using namespace std;

vector<int> solution(vector<string> gems) {
    set<string> getGemTypeCnt;
    for (string gem : gems)
        getGemTypeCnt.insert(gem);
    
    int typeCnt = getGemTypeCnt.size();
    
    unordered_map<string, int> cntGems;
    set<string> types;
    
    int start = 0, end = 0;
    int minRange = 100'000, minEnd;
    
    while (end < gems.size()) {
        cntGems[gems[end]]++;
        types.insert(gems[end]);
    
        if (types.size() == typeCnt) {
            while (start < end) {
                if (cntGems[gems[start]] > 1) {
                    cntGems[gems[start]]--;
                    start++;
                }
                else break;
            }
            if (minRange > end - start) {
                minRange = end - start;
                minEnd = end;
            }
        }
        end++;
    }
    return {minEnd - minRange + 1, minEnd + 1};
}

File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.java
import java.util.*;

class Solution {
    public int[] solution(String[] gems) {
        int size = gems.length;
        
        Set<String> types = new HashSet<>();
        Map<String, Integer> cntGems = new HashMap<>(); 
        
        for (String gem : gems) {
            types.add(gem);
            cntGems.put(gem, 0);
        }
        int typeCount = types.size();
        
        Set<String> cntTypes = new HashSet<>();
        int start = 0, end = 0, minLen = Integer.MAX_VALUE, minEnd = 0;
        for (String gem : gems) {
            cntTypes.add(gem);
            cntGems.put(gem, cntGems.get(gem) + 1);
            
            if (cntTypes.size() == typeCount) {
                while (start <= end) {
                    int nowGemCount = cntGems.get(gems[start]);
                    if (nowGemCount > 1)
                        cntGems.put(gems[start], nowGemCount - 1);

                    else {
                        if (end - start < minLen) {
                            minLen = end - start;
                            minEnd = end;
                        }
                        break;
                    }
                    start++;
                }
            }
            end++;

        }
        return new int[]{minEnd - minLen + 1, minEnd + 1};
    }
}